event-sourcing/event-source-api/src/test/java/uk/gov/justice/services/eventsourcing/source/api/feed/common/EventsPageServiceTest.java

```diff
@@ -0,0 +1,284 @@
+
+        assertThat(paging.getFirst().toString(), is("http://server:123/context/event-streams/" + streamId + "/" + FIRST.getPosition() + "/" + FORWARD + "/" + 2L));
+
+    }
+
+
+    @Test
+    public void shouldReturnFeedWhenSameNumberOfRecordsAsPageSizeWhenLookingForOlderEvents() throws Exception {
+        final UUID streamId = randomUUID();
+
+        final List<EventEntry> events = new ArrayList<>();
+
+        final JsonObject payloadEvent4 = createObjectBuilder().add("field4", "value4").build();
+        final JsonObject payloadEvent3 = createObjectBuilder().add("field3", "value3").build();
+
+        final EventEntry event4 = new EventEntry(randomUUID(), streamId, 4L, "Test Name4", payloadEvent4, new UtcClock().now().toString());
+        final EventEntry event3 = new EventEntry(randomUUID(), streamId, 3L, "Test Name3", payloadEvent3, new UtcClock().now().toString());
+
+        events.add(event4);
+        events.add(event3);
+
+        when(service.recordExists(5L, streamId)).thenReturn(false);
+
+        when(service.recordExists(2L, streamId)).thenReturn(true);
+
+        when(service.events(argThat(new PositionArgumentMatcher(4L)), eq(BACKWARD), eq(2L), eq(streamId))).thenReturn(events);
+
+        final ResteasyUriInfo uriInfo = new ResteasyUriInfo(create("http://server:123/context/"), create("event-streams/" + streamId));
+
+        final Page<EventEntry> feedActual = eventsPageService.pageEvents("4", BACKWARD, 2L, uriInfo, streamId);
+
+        final List<EventEntry> feed = feedActual.getData();
+
+        final PagingLinks paging = feedActual.getPagingLinks();
+
+        assertThat(feed, hasSize(2));
+        assertThat(feed.get(0).getStreamId(), is(streamId.toString()));
+        assertThat(feed.get(0).getSequenceId(), is(4L));
+        assertThat(feed.get(0).getPayload(), is(payloadEvent4));
+        assertThat(feed.get(1).getStreamId(), is(streamId.toString()));
+        assertThat(feed.get(1).getSequenceId(), is(3L));
+        assertThat(feed.get(1).getPayload(), is(payloadEvent3));
+
+        assertThat(paging.getPrevious(), is(nullValue()));
+
+        assertThat(paging.getNext().toString(), is("http://server:123/context/event-streams/" + streamId + "/" + 2 + "/" + BACKWARD + "/" + 2L));
+
+        assertThat(paging.getHead().toString(), is("http://server:123/context/event-streams/" + streamId + "/" + HEAD + "/" + BACKWARD + "/" + 2L));
+
+        assertThat(paging.getFirst().toString(), is("http://server:123/context/event-streams/" + streamId + "/" + FIRST.getPosition() + "/" + FORWARD + "/" + 2L));
+    }
+
+    @Test
+    public void shouldReturnLinkForPreviousPageIfOnLast() throws Exception {
+
+        final UUID streamId = randomUUID();
+
+        final List<EventEntry> events = new ArrayList<>();
+
+        final EventEntry event1 = new EventEntry(randomUUID(), streamId, 1L, "Test Name1", createObjectBuilder().add("field1", "value1").build(), new UtcClock().now().toString());
```

Comment:
we could make some private methods. createEventEntry, mockRecordExists etc. to cleanup some of the tests. Make it more obvious which bits are important details